src/inmanta/data/__init__.py

```diff
@@ -402,41 +405,395 @@ def as_end_filter(
+        col_name = self.order_by_column
+        value = coll_type.get_value(column_value)
+
+        ac = ArgumentCollector(offset=offset - 1)
+        filter = f"{coll_type.get_accessor(col_name)} {relation} {ac(value)}"
+        return [filter], ac.args
+
+    def get_order_elements(self, invert: bool) -> List[Tuple[ColumnNameStr, ColumnType, PagingOrder]]:
+        """
+        return a list of column/column type/order triples, to format an ORDER BY or FILTER statement
+        """
+        order = self.get_order(invert)
+        return [
+            (self.order_by_column, self.get_valid_sort_columns()[self.order_by_column], order),
+        ]
+
+    def get_order_by_statement(self, invert: bool = False, table: Optional[str] = None) -> str:
+        """Return the actual order by statement, as derived from get_order_elements"""
+        order_by_part = ", ".join(
+            (f"{type.get_accessor(col, table)} {order.db_form}" for col, type, order in self.get_order_elements(invert))
+        )
+        return f" ORDER BY {order_by_part}"
+
+    def get_paging_boundaries(self, first: dict[str, object], last: dict[str, object]) -> PagingBoundaries:
+        """Return the page boundaries, given the first and last record returned"""
+        if self.get_order() == PagingOrder.ASC:
+            first, last = last, first
+
+        order_column_name = self.order_by_column
+        order_type: ColumnType = self.get_valid_sort_columns()[self.order_by_column]
+
+        def assert_not_null(in_value: Optional[PRIMITIVE_SQL_TYPES]) -> PRIMITIVE_SQL_TYPES:
+            # Make mypy happy
+            assert in_value is not None
+            return in_value
+
+        return PagingBoundaries(
+            start=assert_not_null(order_type.get_value(first[order_column_name])),
+            first_id=None,
+            end=assert_not_null(order_type.get_value(last[order_column_name])),
+            last_id=None,
+        )
+
+
+class AbstractDatabaseOrderV2(SingleDatabaseOrder, ABC):
+    """
+    Abstract Base class for ordering when using
+    - a user specified order
+    - an additional built in order to make the ordering unique (the id_collumn)
+    """
 
     @property
-    def id_column(self) -> ColumnNameStr:
+    @abstractmethod
+    def id_column(self) -> Tuple[ColumnNameStr, ColumnType]:
+        """Name and type of the id column of this database order"""
+        pass
+
+    # External API
+    def as_filter(
+        self,
+        offset: int,
+        column_value: Optional[PRIMITIVE_SQL_TYPES] = None,
+        id_value: Optional[PRIMITIVE_SQL_TYPES] = None,
+        start: Optional[bool] = True,
+    ) -> Tuple[List[str], List[object]]:
+        """
+        Produce a filter for this order, to select all record before or after the given id
+
+        :param offset: the next free number to use for query parameters
+        :param column_value: the value for the user specified order
```

Comment:
From the coalescing in the filter builder I actually get the impression we do try to support filtering on nullable columns, in which case shouldn't we treat `None` for the column (and id?) values as `null` rather than absence of a filter? And then why do we `assert_not_null`?

Reply:
Does this method expect coalesced arguments? If this is the case I think it should be part of the docstring.

Reply:
My confusion about `null` vs unspecified applies to `RequestedPagingBoundaries` as well.